File: my-code/1405/main.go
package main

import (
	"fmt"
	"sort"
)

// @Time     : 2022/2/7 10:00
// @Author   : rs
// @File     : main.go

func longestDiverseString(a int, b int, c int) string {
	var ans []byte
	cnt := []struct {
		c  int
		ch byte
	}{
		{a, 'a'},
		{b, 'b'},
		{c, 'c'},
	}
	for {
		sort.Slice(cnt, func(i, j int) bool {
			return cnt[i].c > cnt[j].c
		})
		hasNext := false
		for i, p := range cnt {
			if p.c <= 0 {
				break
			}
			m := len(ans)
			if m >= 2 && ans[m-2] == p.ch && ans[m-1] == p.ch {
				continue
			}
			hasNext = true
			ans = append(ans, p.ch)
			cnt[i].c--
			break
		}
		if !hasNext {
			return string(ans)
		}
	}
}

func main() {
	fmt.Println(longestDiverseString(1, 1, 7))
}

File: my-code/1405/main.py
# -*- coding:utf-8 -*-
# @Time     : 2022/2/7 09:50
# @Author   : rs
# @File     : main.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        ans = []
        cnt = [[a, 'a'], [b, 'b'], [c, 'c']]
        while True:
            cnt.sort(key=lambda x: -x[0])
            has_next = False
            for i, (c, ch) in enumerate(cnt):
                if c <= 0:
                    break
                if len(ans) >= 2 and ans[-2] == ch and ans[-1] == ch:
                    continue
                has_next = True
                ans.append(ch)
                cnt[i][0] -= 1
                break
            if not has_next:
                return ''.join(ans)


if __name__ == '__main__':
    print(Solution().longestDiverseString(a=1, b=1, c=7))
